api/agent/drivers/docker/docker_networks.go

```diff
@@ -0,0 +1,194 @@
+package docker
+
+import (
+	"context"
+	"math"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/fnproject/fn/api/agent/drivers"
+	"github.com/fnproject/fn/api/common"
+	"github.com/sirupsen/logrus"
+	"golang.org/x/time/rate"
+)
+
+type DockerNetworks struct {
+	// protects networks map
+	networksLock sync.Mutex
+	networks     map[string]uint64
+
+	// read-only map for unlocked reads for WaitDockerNetworks
+	waitList      map[string]struct{}
+	waitListReady chan struct{}
+}
+
+func NewDockerNetworks(conf drivers.Config) *DockerNetworks {
+	obj := &DockerNetworks{
+		networks:      make(map[string]uint64),
+		waitList:      make(map[string]struct{}),
+		waitListReady: make(chan struct{}),
+	}
+
+	// Record DockerNetworks both in wait-list and in network allocations
+	for _, net := range strings.Fields(conf.DockerNetworks) {
+		obj.networks[net] = 0
+		obj.waitList[net] = struct{}{}
+	}
+
+	// IMPORTANT: if PreForkPool is enabled with additional networks, add those in wait-list
+	if conf.PreForkPoolSize != 0 {
+		for _, net := range strings.Fields(conf.PreForkNetworks) {
+			obj.waitList[net] = struct{}{}
+		}
+	}
+
+	return obj
+}
+
+func (n *DockerNetworks) isDockerNetworkReady() bool {
+	if len(n.waitList) > 0 {
+		select {
+		case <-n.waitListReady:
+		default:
+			return false
+		}
+	}
+	return true
```

Comment:
hmm, i figured part of the use of this patch was to wait for multiple networks. not saying to do that, just seems odd to need all this machinery in that case. maybe the wrong place to thread, but I find:

> In addition, if driver.IsNetworkReady() is false, then status call is executed using network disabled. This ensures that Status call runs successfully while waiting for docker networks.

this seems questionable. we want status calls to run successfully for some other info presumably (no reason listed, $jumping_to_conclusions)? but aside from seeming odd it seems like a pretty good stand in for exactly what this patch does in a simpler way, that is, if the docker networks aren't ready then status calls will fail, right? at least intuitively, this seems desirable. is there information from a status call that the lb or another component would use to ever route traffic to this node before the networks are ready? is there another reason (to satisfy some health checker condition that would potentially take out a node that is in fact starting up? something else?)?

Reply:
additionally quoted condition seems troublesome for reasoning about. i'm sure things are too far gone as it is (for easily reasoning about... health), but it is a bit hairy to go flipping 'oh the status call succeeds when networking is off, then gets flipped on and the networking starts failing so the status call starts failing' -- seems odd to change the status call in place and let it succeed in the first place, makes sense to just fail i guess. one more thing i guess

Reply:
I agree, we can simply let the status fail. This PR lets status succeed to exercise more code paths in docker and agent before the networks are available, but at the cost of this additional complexity.